src/selectors/selectExplorerTree.ts

```diff
@@ -0,0 +1,223 @@
+import platformPath from 'path';
+import { Uri } from 'vscode';
+import { CaseHash } from '../cases/types';
+import { caseAdapter, jobAdapter, State } from '../data/slice';
+import { Job, JobHash, JobKind, PersistedJob } from '../jobs/types';
+import { LeftRightHashSetManager } from '../leftRightHashes/leftRightHashSetManager';
+import { buildHash, isNeitherNullNorUndefined } from '../utilities';
+
+export const doesJobAddNewFile = (kind: Job['kind']): boolean => {
+	return [
+		JobKind.copyFile,
+		JobKind.createFile,
+		JobKind.moveAndRewriteFile,
+		JobKind.moveFile,
+	].includes(kind);
+};
+
+export type ExplorerNodeHashDigest = string & {
+	__ExplorerNodeHashDigest: 'ExplorerNodeHashDigest';
+};
+
+export type ExplorerNode =
+	| Readonly<{
+			hashDigest: ExplorerNodeHashDigest;
+			kind: 'ROOT';
+			label: string;
+	  }>
+	| Readonly<{
+			hashDigest: ExplorerNodeHashDigest;
+			kind: 'TOP';
+			label: string;
+	  }>
+	| Readonly<{
+			hashDigest: ExplorerNodeHashDigest;
+			kind: 'DIRECTORY';
+			label: string;
+	  }>
+	| Readonly<{
+			hashDigest: ExplorerNodeHashDigest;
+			kind: 'FILE';
+			label: string;
+			jobHash: string;
+			fileAdded: boolean;
+	  }>;
+
+export const buildRootNode = (): ExplorerNode & { kind: 'ROOT' } => ({
+	hashDigest: buildHash('ROOT') as ExplorerNodeHashDigest,
+	kind: 'ROOT',
+	label: '',
+});
+
+export const buildTopNode = (
+	label: string,
+): ExplorerNode & { kind: 'TOP' } => ({
+	hashDigest: buildHash('TOP') as ExplorerNodeHashDigest,
+	kind: 'TOP',
+	label,
+});
+
+export const buildDirectoryNode = (
+	path: string,
+	label: string,
+): ExplorerNode & { kind: 'DIRECTORY' } => ({
+	hashDigest: buildHash(
+		['DIRECTORY', path, label].join(''),
+	) as ExplorerNodeHashDigest,
+	kind: 'DIRECTORY',
+	label,
+});
+
+export const buildFileNode = (
+	job: PersistedJob,
+	label: string,
+): ExplorerNode & { kind: 'FILE' } => {
+	const fileAdded = doesJobAddNewFile(job.kind);
+
+	return {
+		kind: 'FILE',
+		hashDigest: buildHash(
+			['FILE', job.hash, label].join(''),
+		) as ExplorerNodeHashDigest,
+		label,
+		jobHash: job.hash,
+		fileAdded,
+	};
+};
+
+export type ExplorerTree = Readonly<{
+	rootNodeHashDigest: ExplorerNodeHashDigest;
+	nodes: Record<ExplorerNodeHashDigest, ExplorerNode>;
+	children: Record<
+		ExplorerNodeHashDigest,
+		ReadonlyArray<ExplorerNodeHashDigest>
+	>;
+	parents: Record<ExplorerNodeHashDigest, ExplorerNodeHashDigest>;
+	selectedNodeHashDigest: ExplorerNodeHashDigest | null;
+	expandedNodeHashDigests: ReadonlyArray<ExplorerNodeHashDigest>;
+	appliedJobHashes: ReadonlyArray<JobHash>;
```

Comment:
do we want to keep domain specific field `appliedJobHashes` in tree structure or move it to separate structure?

Reply:
maybe lets rename it to something like `selectedTreeNodes` (common field for all trees that have checkboxes and can be selected) or move it to other structure

Reply:
selectedTreeNodes are already a used name (selectedNodeHashDigest), the concept of selection (focus?) is the concept for the tree

Reply:
maybe then lets rename `selectedNodeHashDigest ` to `focusedNodeHashDigest` and use name `selectedNodeHashDigests` to store nodes that are selected using checkboxes?